compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1506,13 +1503,36 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         });
 
         let existential_projections = bounds.projection_bounds.iter().map(|(bound, _)| {
-            bound.map_bound(|b| {
-                if b.projection_ty.self_ty() != dummy_self {
-                    tcx.sess.delay_span_bug(
-                        DUMMY_SP,
-                        &format!("trait_ref_to_existential called on {:?} with non-dummy Self", b),
-                    );
+            bound.map_bound(|mut b| {
+                assert_eq!(b.projection_ty.self_ty(), dummy_self);
+
+                // Like for trait refs, verify that `dummy_self` did not leak inside default type
+                // parameters.
+                let references_self = b.projection_ty.substs.iter().skip(1).any(|arg| {
+                    if let ty::GenericArgKind::Type(ty) = arg.unpack() {
```

Comment:
this feels incorrect for const generics. What happens if you have a generic constant referencing self? (using `feature(generic_const_exprs)`)

a generic constant not using self would be even worse :sweat_smile: because it still has `Self` in its generic arguments even if unused.

Reply:
were there any problems with using `arg.walk` instead of `ty.walk`. Also, the generic arg itself should get returned as the first item of `walk`, so the explicit `ty == dummy_self` check shouldn't be necessary

Reply:
use `feature(generic_const_exprs)` and have a `Trait<{ N + 1 }` in it, or even `Trait<{ 1 + 2 }>`.

can look into making a full example if you want, am done for today :sweat_smile: